students/erik_oien/lesson05/mailroom3.py

```diff
@@ -0,0 +1,165 @@
+# create report menu
+
+def create_report():
+    while True:
+        donor_report = []
+        [donor_report.append(donor_info(donor)) for donor in donor_dict.items()]
+        donor_report_info = sorted(donor_report, key=sort_key, reverse=True)
+        table_header()
+        [row_formatter(donor) for donor in donor_report_info]
+        return_input = input(f"{sub_menu_prompt}")
+        if return_input in ["menu", "quit"]:
+            sub_menu_switch[return_input.lower()]()
+            return False
+        else:
+            print("please enter a valid option")
+        
+# send letters to all donors
+
+def thank_all_donors():
+    while True:
+        wd = os.getcwd()
+        dir_name = input(f"What would you like to call the directory?\n{sub_menu_prompt}")
+        if dir_name.lower() in ["menu", "quit"]:
+            sub_menu_switch[dir_name.lower()]()
+            return False
+        else:
+            try:
+                dir_path = wd + "/" + dir_name
+                os.mkdir(wd + "/" + dir_name)
+                for donor in donor_dict.items():
+                    donor_facts = donor_info(donor, all_donors=True)
+                    donor_letter = thank_you_letter(donor_facts, dir_path)
+                return False
+            except (FileExistsError):
+                print("Please try a different file name")
+
+# helper functions
+
+# function to create email
+def thank_you_email(donor, amount):
+    email_text = "Dear {},\n\nThank you for your generous donation of ${}! As you certianly know, kittens\nand metal are awesome and your donation will insure that others will be able\nto enjoy kittens and metal.\n\nSincerely,\n\nKitten and Metal Charity \m/\n\n"
+    return email_text.format(donor, amount)
+
+# function for table header
+def table_header():
+    print("{:25}|{:12}|{:10}|{:12}".format("Donor Name", "Total Given", "Num Gifts", "Average Gift"))
+    print("-" * 62)
+
+# function to format rows
+def row_formatter(row):
+    print("{:25}|${:11.2f}|{:10}|${:11.2f}".format(*row))
+
+# function to quit program
+def quit_program():
+    print("Bye!")
+    sys.exit()  
+
+# sort key
+def sort_key(donor_report):
+        return donor_report[1]
+
+# donor info:
+
+def donor_info(donor, all_donors=False):
+    donor_name = donor[1]["first_name"] + " " + donor[1]["last_name"]
+    donor_sum = sum(donor[1]["donations"])
+    donor_length = len(donor[1]["donations"])
+    donor_average = donor_sum / donor_length
+    if all_donors:
+        return(donor_name, donor_length, donor_sum)
+    else:
+        return(donor_name, donor_sum, donor_length, donor_average)
+
+# thank you letter
+def thank_you_letter(text, dir_path):
+    letter_text = "Dear {},\n\n\tYou have made {} donations to the kittens and metal charity totaling ${}.\n\nSincerely,\n\n\tKitten and Metal Charity \m/".format(*text)
+    file_name = dir_path + "/" + text[0] + ".txt"
+    letter = open(file_name, "w")
+    letter.write(letter_text)
+    letter.close()
```

Comment:
best practices is to use `with` statement which handles closing the file for you.